4-Beyond-Classical-Search/c_hillClimbing.js

```diff
@@ -1,144 +1,166 @@
-      this.blockWidth = this.hillDiagram.blockWidth;
-      this.botHeight = 50;
-      this.botWidth = this.blockWidth + 20;
-      this.xOffset = 10;
-      this.yOffset = 40;
-      this.renderHillClimber();
-    }
-    renderHillClimber() {
-      let robotLocation = this.hillClimber.getCurrentState();
-      let robotStateValue = this.states[robotLocation];
-      this.robot = this.svg.append('g')
-        .attr('class', 'robot')
-        .append('svg:image')
-        .attr('xlink:href', '../third-party/robot.png')
-        .attr('height', this.botHeight)
-        .attr('width', this.botWidth)
-        .attr('x', this.xScale(robotLocation) - this.xOffset)
-        .attr('y', this.h - this.yScale(robotStateValue) - this.yOffset);
-      this.hillDiagram.visit(robotLocation);
+//Class to draw the robot
+class HillClimberDiagram {
+  constructor(hill, svg, h, w, hillDiagram, hillClimber) {
+    this.h = h;
+    this.w = w;
+    this.hill = hill;
+    this.hillDiagram = hillDiagram;
+    this.hillClimber = hillClimber;
+    this.states = this.hill.getStates();
+    this.svg = svg;
+    this.xScale = this.hillDiagram.xScale;
+    this.yScale = this.hillDiagram.yScale;
+    this.blockWidth = this.hillDiagram.blockWidth;
+    this.botHeight = 50;
+    this.botWidth = this.blockWidth + 20;
+    this.xOffset = 10;
+    this.yOffset = 40;
+    this.renderHillClimber();
+  }
+  renderHillClimber() {
+    let robotLocation = this.hillClimber.getCurrentState();
+    let robotStateValue = this.states[robotLocation];
+    this.robot = this.svg.append('g')
+      .attr('class', 'robot')
+      .append('svg:image')
+      .attr('xlink:href', '../third-party/robot.png')
+      .attr('height', this.botHeight)
+      .attr('width', this.botWidth)
+      .attr('x', this.xScale(robotLocation) - this.xOffset)
+      .attr('y', this.h - this.yScale(robotStateValue) - this.yOffset);
+    this.hillDiagram.visit(robotLocation);
 
-    }
-    move(state) {
-      let robotLocation = state;
-      let robotStateValue = this.states[state];
-      this.robot.transition()
-        .duration(100)
-        .attr('x', this.xScale(robotLocation) - this.xOffset)
-        .attr('y', this.h - this.yScale(robotStateValue) - this.yOffset);
-    }
   }
+  move(state) {
+    let robotLocation = state;
+    let robotStateValue = this.states[state];
+    this.robot.transition()
+      .duration(100)
+      .attr('x', this.xScale(robotLocation) - this.xOffset)
+      .attr('y', this.h - this.yScale(robotStateValue) - this.yOffset);
+  }
+}
 
-  //Wrapper class for the entire diagram
-  class HillClimbingDiagram {
-    constructor(selector, h, w) {
-      this.h = h;
-      this.w = w;
-      this.svg = d3.select(selector).html("")
-        .append('svg')
-        .attr('height', this.h)
-        .attr('width', this.w)
-        .attr('border', 2);
+class HillWorld {
+  constructor(selector, h, w) {
+    this.h = h;
+    this.w = w;
+    this.svg = d3.select(selector).html("")
```

Comment:
.html("") is used here to delete the previously added elements. It is applicable when 'restart' is pressed and a new instance of the class is created.